### ISSUE TITLE
Document collections

### ISSUE THREAD

**Issue by simonster** [zotero/zotero#21]
Make items cited in open documents appear on the left side of the Zotero pane.

**Comment by aurimasv** [comment]
I'd like to put some effort into this. Could we come up with a consensus on what the UX for this should be? @fbennett's implementation is a step in the right direction, but I think it needs to be a lot more streamlined. Here's my take on it:

* Because documents can contain items from multiple libraries (multiple personal as well as group libraries), they should be displayed in an independent category (let's call it "Open Documents"), which would similar to the "Group Libraries" category. In this category, you could have a number of currently-open documents, which would appear as libraries. No collections are involved (at least at this point). Clicking on each library (or open document) would display the list of items in the document.
* These libraries are created (and removed) dynamically. They are created when a new document is opened and some Zotero operation is performed from within Word. Zotero obviously has no way of knowing when a new document is opened, hence the need to perform some Zotero-related action (e.g. Refresh, but most actions scan the whole document anyway). It may be possible to automatically trigger this from the Zotero Word Processor add-in, but given how slow these operations can be on larger documents, I don't think we would want to bog down opening documents.
* The items are populated in the following order:
  * by looking up the reference URI: if the reference is already in the user's library, the item is just a link to that item
  * by extracting embedded metadata
  * by looking up metadata on zotero.org (if the item is from a public library)
  * by randomly generating metadata. jk, I'm not really sure what to do if the metadata is not embedded in the document and only a URI is available to some private library. Maybe not display that item at all? Maybe display some sort of placeholder based on rendered citation?
* The items are updated (from document to Zotero) only when operations are performed on the document that involve fetching all items from the document. I think this pretty much means that it will stay in-sync with the document unless a reference is deleted. In the future, we may want to think how to streamline things for users that attempt to edit citations within the document.
* Since items in the "Open Document" library are links to actual documents in the user library (where available), editing those items in the "Open Document" library would change metadata of the linked item. Items that are not part of user-accessible libraries, would be read-only (perhaps with some indicator showing that the item belongs to some other library). That's sort of a stopgap measure for now, because I think that we would just want users to be able to edit these references as well, but basically save them in their library (where?) and re-insert the reference.

That's all I've got for now. I'm sure I'm missing something.

One related feature that we should consider is converting all references in a document to user or group library references. There might be an issue about that somewhere already.

**Comment by fbennett** [comment]
This would be great! I'm sure it can be done more elegantly.

On the last bullet-point, you could allow a user to drag the virtual item to a library where they have write access, add the URI to it, and update the document with the amended copy of the reference. It would introduce an additional wrinkle, though. In the MLZ group-extraction approach, extracted refs all end up in a library shared among the researchers working on the document. The new ref becomes the "canonical" copy for the document.

If dragging to arbitrary libraries is allowed, two researchers working on the same document might each drag it to locations accessible only to themselves. I think you would want a date-stamp on the item data in the document, so that library data can be synced to it without exceptional overhead. There would still be a potential privacy issues for items dragged into a personal library, if their association with an ongoing external document project is later forgotten.

**Comment by bwiernik** [comment]
For copying items to another library, I think it might be good to use a UI other than dragging. A button in the right pane that says "Copy to Library" that opens a selection dialogue to choose a library. This pane can include a warning that the library needs to be accessible to all authors. Using this copy function should change the canonical version of the item in the document to the copied version. 

For items already in a library, the same button could say "Show item in < Library Name>. Using a button instead of dragging removes a variety of potentially confusing problems:
- What happens when you drag an item to two different libraries? 
- Users might try to add an item to a document by dragging items to the Document collection. 

For items without embedded metadata, I think a placeholder with a message saying that metadata is not stored would be least confusing.

**Comment by fbennett** [comment]
[sorry about the clutter here, looking forward to document collections when they arrive]

**Comment by rmzelle** [comment]
Shameless plug for people reading this issue: as long as document collections aren't implemented in Zotero itself, users can use http://rintze.zelle.me/ref-extractor/ (code at https://github.com/rmzelle/ref-extractor/) as a workaround. Currently ref-extractor only allows you to extract citation metadata from Word .docx files. You can save this metadata in CSL JSON format and then import it into Zotero as a document collection.

**Comment by adomasven** [comment]
https://github.com/zotero/zotero/pull/1242 also contained a [commit](https://github.com/zotero/zotero/pull/1242/commits/2ad0dc00daa945f22041612639abfcb001b96408) that adds some important functionality to allow implementing document collections.

Namely, the items in the last document that was active with Zotero are now accessible under

```
Zotero.Integration.currentSession.getItems()
```

**Comment by fbennett** [comment]
To take a step back from the ignorant and confused gushing of my earlier post, I wonder whether there is a point to document collections as described. Users seem to want one of two things: to recover references from a document and stick them in a collection that they own; or to push document references into a shared space where project collaborators all have library access. Both operations simplify things for the user by putting items in one place and giving them control. A mixed view of items, some local, some from libraries in various permission states, some reconstructed from document data, all in various states of linkage to other items elsewhere, that's interesting at a technical level, but it seems to cut against the aim of streamlining things.

**Comment by bwiernik** [comment]
@fbennett The most common application for them I see is a desire to simply see all of the items that have been cited in a document (e.g., to check for missing citations or choose which ones to eliminate).

**Comment by fbennett** [comment]
Yes, but extracting to a user-owned collection, with creation of items that are not found in their library, would do that. Same for extract-to-group-library. What is gained by showing the user a container with items in an unfamiliar mixture of states?

**Comment by dstillman** [comment]
Just having a virtual collection show up when you have a document open seems like the most desirable approach here from a user's perspective. That would let you copy missing items to your library and also enable various other things, like quickly exporting a set of items that were used in a given document or making edits in one place without needing to search or keep a static collection updated. Certainly harder technically than extraction, but I think it'd be a pretty valuable feature.

**Comment by retorquere** [comment]
Could a pseudo-search maybe be a vehicle to achieve this?

**Comment by adomasven** [comment]
I'm thinking more about the capabilities that we would want this to have.

We would call this something like "Document Explorer" as a button in the Zotero plugin interface. It opens up a new window with 2 lists. The two lists are citations in the document, and items in the document. The following functionality is available:

- Find all citations that cite a specific item
- List items in a selected citation
- Highlight a selected citation in the document
- Open up the citation dialog for the selected citation
- Edit item metadata
- Highlight unlinked items and allow relinking them (requires yet another dialog?)
- Highlight items that are only cited in the bibliography
- Show item in Zotero library

Concerns:
1. Being able to highlight citations in a document from a list is a super useful and one of the core functionalities for a dialog like this, but also one that would likely be used primarily to remove those citations. It will however break the citation highlighting and citation dialog functionality if citations are removed while the document explorer dialog is open. Getting fields from the document is an expensive operation, so we cannot be continuously running it in the background. We could add a "Reload" button for advanced users who know what they are doing. Maybe we could check by running the `cursorInField()` integration call, which returns the field data for the cursor in field and compare it to expectation, but that would then make these operations much slower and also users would see us intermediately highlighting the wrong item.
2. Unlinked items are also items that your collaborator cited from their own library, so this might create confusion for collaborative writing. We should insist that users relink unlinked citations that got unlinked accidentally, but balance it with not making them relink citations that are linked to someone elses library. On the other hand, maybe this is not a big concern in practice.
